src/libdredd/src/mutation_replace_binary_operator.cc

```diff
@@ -269,10 +330,20 @@ std::string MutationReplaceBinaryOperator::GenerateMutatorFunction(
     mutant_offset++;
   }
 
-  new_function
-      << "  return " << arg1_evaluated << " "
-      << clang::BinaryOperator::getOpcodeStr(binary_operator_.getOpcode()).str()
-      << " " << arg2_evaluated << ";\n";
+  const clang::BuiltinType* exprType =
+      binary_operator_.getType()->getAs<clang::BuiltinType>();
+  if (exprType->isBooleanType() ||
+      (exprType->isInteger() && !exprType->isBooleanType() &&
+       !binary_operator_.isLValue())) {
+    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                 << mutant_offset << ")) return !(" << getExpr(ast_context)
+                 << ");\n";
+    mutant_offset++;
+  }
+
+  GenerateConstantInsertion(new_function, exprType, mutant_offset);
```

Comment:
I feel like it might be clearer to also extract the code for operator replacement and unary operator insertion into functions, so that you'd have:

```
GenerateBinaryOperatorReplacement(...);
GenerateArgumentReplacement(...);
GenerateUnaryOperatorInsertion(...);
GenerateConstantReplacement(...);
```

The first would be all the logic for replacing one binary operator with another. The second the logic for replacing the operator with one of its arguments. The third the logic for inserting a unary operator. And the fourth would be replacing with a constant.

As mentioned below, I actually think `GenerateUnaryOperatorInsertion` and `GenerateConstantReplacement` could be static methods of `MutationReplaceExpr`, which get invoked from all three classes as I guess the logic should be the same.

BTW: in writing this comment it occurred to me that "ConstantInsertion" isn't really the right words, as it's a replacement, not an insertion.

Reply:
I have done the refactoring described above with `GenerateUnaryOperatorInsertion` and `GenerateConstantReplacement` as static methods in `MutationReplaceExpr`.